README.md
# node-image-farmer

*"node-image-farmer...because it crops images"*

* Original connect-thumbs code by [Irakli Nadareishvili](https://github.com/inadarei), modified by [Allan Bogh](https://github.com/ajbogh)

Image thumbnailing middleware for Connect.js/Express.js that integrates with content-aware
cropping provided by [Smartcrop.js](https://github.com/jwagner/smartcrop.js/)

node-image-farmer implements the boilerplate code for creating thumbnails of large images in a standard,
Connect.js-complient way, allowing sensible defaults and high degree of customization.

## Installation

### Dependencies

C++11 compiler - g++ 4.8 or clang 3.2
Why? Canvas doesn't install using node-gyp without C++11 features. 
Sometimes the root user can't perform the npm install, however a normal user account can without installing g++ 4.8. 

### Using your own API

See app/app.js for an example API.

```console
$ # dependencies need to be installed first
$ npm install node-image-farmer --save
``` 

### Using the provided API

```console
$ git clone git@github.com:ajbogh/node-image-farmer.git
$ cd node-image-farmer
$ npm run install-ubuntu-deps
$ # npm run install-redhat-deps
$ # npm run install-mac-deps (untested)
$ npm install
```
   
### Checking out specific version

```console
$ git fetch --tags
$ git checkout v1.2.1
$ # npm install
```
    
### Installing Dependencies Manually (if "npm run install-XYZ-deps" doesn't work)   

Note: there are also APT and YUM repositories you can use for Ubuntu/Debian and 
RedHat/Centos/Fedora respectively.

On OS-X you can easily install dependencies with: 

```console
> brew install graphicsmagick
> xcode-select --install
> brew install pkgconfig
> brew install pixman
> brew install libjpeg
> brew install giflib 
> brew install cairo
```

## Serving local images

You can serve images from your own server by creating a symlink under app/images

```
$ cd app
$ ln -s images /path/to/your/images/folder
```

You can navigate your images folder similar to how you normally would with a URL:

```
http://localhost:3000/content/smart/small/my/subfolder/myImage.jpg
```

## API

- **preset**: \[full, small, medium, hero, irakli] (default, configurable)
- **width**: w or width (query string, optional)
- **height**: h or height (query string, optional)
- **quality**: q or quality (query string, optional, 1-100 default 95)
- **minScale**: minScale (query string, optional, between 0.1 and 1) - Allows the system to zoom in on a portion of the image to 
  find the best crop setting. Lower numbers may take an extremely long time, we recommend not using numbers below 0.5.

All images by default are served from port 3000 (configurable) and reside in the /content/smart subfolder
```
http://localhost:3000/content/smart/[preset]/[path/to/image]/[image.jpg or ?base64=ab123]?w=123&h=456&q=95&minScale=1.0
```

Presets are defined in the configuration, but default presets are 'irakli', 'small', 'medium', and 'hero'. 
If no preset matches or you would like the full image, you may use 'full' and optionally define width and height parameters.

```
http://localhost:3000/content/smart/small/myImage.jpg
http://localhost:3000/content/smart/medium/myImage.jpg
http://localhost:3000/content/smart/full/myImage.jpg
```

You may override preset crops with the w, width, h, or height parameters. You may also override the quality with the 'q' or 'quality' parameter.

```
http://localhost:3000/content/smart/small/myImage.jpg?w=150&h=100
http://localhost:3000/content/smart/medium/myImage.jpg?width=150&height=100
http://localhost:3000/content/smart/full/myImage.jpg?w=150&height=100&quality=50
http://localhost:3000/content/smart/full/myImage.jpg?q=2
```

## Running an Example

If you have all the prerequisites installed you can launch a demo with:

```
$ npm run app
$ # or
$ ./nodeImageServer
```

And then open your browser at the [following URL](http://localhost:3000/content/smart/full/?base64=aHR0cDovL3d3dy5wdWJsaWNkb21haW5waWN0dXJlcy5uZXQvcGljdHVyZXMvMTAwMDAvdmVsa2EvMTA4MS0xMjQwMzI3MzE3cGMzcS5qcGc=):

```
http://localhost:3000/content/smart/full/?base64=aHR0cDovL3d3dy5wdWJsaWNkb21haW5waWN0dXJlcy5uZXQvcGljdHVyZXMvMTAwMDAvdmVsa2EvMTA4MS0xMjQwMzI3MzE3cGMzcS5qcGc=
```

You can see on the following diagram what simple (on the left), and smart (on the right)
 crops produce compared to the original (center)
 
 ![](https://raw.githubusercontent.com/inadarei/connect-thumbs/master/example/crops-smart.jpg)

Photo Credit: [Andrew Schmidt](http://www.publicdomainpictures.net/view-image.php?image=2514&picture=seagull&large=1) (Public Domain)

### Smart Cropping

200x400 remote image
```
http://localhost:3000/content/smart/full/?w=200&h=400&base64=aHR0cDovL3d3dy5wdWJsaWNkb21haW5waWN0dXJlcy5uZXQvcGljdHVyZXMvMTAwMDAvdmVsa2EvMTA4MS0xMjQwMzI3MzE3cGMzcS5qcGc=
```

200x400 local image
```
http://localhost:3000/content/smart/full/myImage.jpg?w=200&h=400
```
    
## Connect.js/Express.js Usage (Creating your own API)

    See app/app.js for an example. You may add/modify or remove any code within this file to obtain your own API.
    
when configured with defaults, and if you have your node process running at yourdomain.com, a request such as:

    http://yourdomain.com/content/smart/medium/?base64=aHR0cDovL3VwbG9hZC53aWtpbWVkaWEub3JnL3dpa2lwZWRpYS9jb21tb25zLzYvNjYvRWluc3RlaW5fMTkyMV9ieV9GX1NjaG11dHplci5qcGc=
    
will display Einstein's photo from Wikipedia as a width: 300 (and proportionally resized height) thumbnail.

Another example uses the file system to resize a file within the app/images folder. In a production environment this folder would
be a symlink to wherever the images are located.

```
http://localhost:3000/content/smart/irakli/crops-smart.jpg
```

You may also override a preset by supplying a width, height, and quality in the query strings. Each of these are optional and
distinct properties. For instance, by specifying the width you can override the preset's width without affecting the height or quality properties.

```
http://localhost:3000/content/smart/medium/crops-smart.jpg?height=300&width=200&quality=85
http://localhost:3000/content/smart/medium/crops-smart.jpg?h=300&w=200&q=85
```


This is because:
 
1. `/content/smart/medium` in the beginning of the URL instructs the middleware to use default resizing preset named "medium" 
 which corresponds to proportional resizing to width: 300px.
1. the long, somewhat cryptic code is base64-encoded version of the 
 [URL of Einstein's photo on Wikipedia](http://upload.wikimedia.org/wikipedia/commons/6/66/Einstein_1921_by_F_Schmutzer.jpg)

## Configuration

Edit the config/appConfig.json file for your setup. (comments added below are for informational purposes, not to be included in JSON)

```json
    {
        "baseDirectory": "/content/smart",
        "port": 3000,
        "allowedExtensions" : ["png", "jpg"],
        "allowedMimeTypes" : [
            "image/jpeg",
            "image/pjpeg",
            "image/gif",
            "image/png"
        ],
        "tmpDir" : "/tmp/node-image-farmer",
        "browserTTL": 86400, // cache for 24 hours by default
        "tmpCacheTTL": 1800, // 30 minutes by default
        "fullFileTTL": 86400, // refresh the full file copy after 24 hours
        "useMultipleProcesses": true, //Uses all available cores to process long image requests
        "presets": { //all lowercase, one word
            "irakli": {
                "width": 300,
                "height": 520,
                "quality": 90
            },
            "small" : {
                "width": 240,
                "height": 160,
                "quality": 75,
                "minScale": 0.7
            },
            "medium" : {
                "width": 542,
                "height": 386,
                "quality": 85
            },
            "hero" : {
                "width": 980,
                "height": 370,
                "quality": 90
            }
        }
    }
```

where:

 * baseDirectory - The directory in your URL to load the files from. Use '/someDirectory' 
   to render images from  http://localhost:3000/someDirectory for example.
 * port - The port to use. Default is 3000, you might want port 80 or 443 in production.
 * allowedExtensions - An array of allowed extensions (without dots!)
 * allowedMimeTypes - An array of allowed mime types.
 * tmpDir - The directory to store copied images. Defaults to: `/tmp/node-image-farmer`. 
 * browserTTL - The cache length in seconds to keep images in cache on the client browsers (24 hours by default).
 * tmpCacheTTL - The cache length in seconds to keep the cropped images from re-processing the original image. Defaults to 30 minutes.
 * fullFileTTL - The cache length in seconds to keep a copy of the original image. Defaults to 24 hours.
 * useMultipleProcesses - Set true to use multiple cores for image requests (recommended), 
   false will block requests until previous requests have finished.
 * presets - json object describing various image presets. You can indicate width, height and quality 
   level for each. Quality adjusts image compression level and its value ranges from 0 to 100 (best).

## Serving Behind a Web Server
    
*ATTENTION*: in typical web setups, static content such as images is often served by a web-server, never allowing 
requests to *.jpg, *.png etc. to reach Node process. If you want to use node-image-farmer, obviously you must allow
paths to thumbnailed images to pass through to Node. Please add appropriate exception to you web server configuration. 
For Nginx, your configuration may look something like the following:

```
  # Thumbnail processing
  location ^~ /content/smart {
    auth_basic off;

    proxy_pass         http://127.0.0.1:3000;
    proxy_set_header   Host                   $http_host;
    proxy_redirect off;
  }

  #  static content
  location ~* ^.+.(jpg|jpeg|gif|css|png|js|ico|xml)$ {
    # access_log        off;
    expires           15d;
  }
```

Alternatively, sometimes connect-static is used to serve static content. If you do that, please make sure that 
connect-static fires *after* node-image-farmer does.

## Performance and Scalability

Node.js is very fast, but GraphicsMagick and over-the-HTTP fetching of the original image most certainly are not. In any 
production setup it is highly recommended to put thumbnailing of images behind some sort of proxy/cache/CDN. 

The tmp file storage will prevent the system from needing to crop the image if it has already done it once before (within the time limit). 
The full file cache will reduce the need to download the file from a website or copy the file from a disk.

When useMultipleProcesses is enabled, the system can crop as many images at a time as there are cores on the server. 
It's best to use more than 2 cores, preferably 4. If there is only 1 core then the system will still create 1 new process per request, but the 
OS will handle multithreading.

If useMultipleProcesses is disabled then the process will block new requests until the previous image is cropped and served.

Viable options include:

- Enabling the integrated disk-based cache provided by node-image-farmer. You can do this by passing custom `tmpCacheTTL`
configuration variable when initializing Thumbs. This variable is set in seconds and is 0 by default. Setting it 
to values greater than 0 enables caching.
- Put [Varnish](https://www.varnish-cache.org/) in front of the thumbnail URLs
- Use a robust CDN such as [Amazon's CloudFront](http://aws.amazon.com/cloudfront/)
- Pick your own poison.

## License

[MIT](LICENSE)


CHANGELOG.md
### v1.2.7

- Modified console.logs to use debug instead. Now debugging can be enabled with `DEBUG=node-image-farmer* npm run app`

### v1.2.6

- Modified the run script to include the imageFarmer process name. Now `sudo killall imageFarmer` will work.

### v1.2.5

- Added CHANGELOG.md and adjusted the README.md file.
- Modified nodeImageFarmer.sh to log to /var/log/node-image-farmer and nohup the process.

### v1.2.4

- Set nodeImageFarmer.sh to executable

### v1.2.3

- Added nodeImageFarmer.sh so that the app can be found more easily in `ps -A | grep nodeImageFarmer`

### v1.2.2

- Changed README.md based on fixes from v1.2.1

### v1.2.1

- Fixed the prefix setting so the temp files are written to the correct folder.

### v1.2.0

- Moved the appConfig into a JSON file. Bumped the minor version because it's a semi-breaking change.

### v1.1.0

- Massive refactor of connect-thumbs to use promises, multiple processes, and API changes. 
- Decided to bump the major version to prevent conflicts with older versions of the code. 


lib/node-image-farmer_old.js
if (!module.parent) { console.log("Please don't call me directly. I am just the main app's minion."); process.exit(1); }

// example:
// http://example.com/thumbs/small/hashcode.jpeg

var ttl
  , tmpCacheTTL
  , tmpDir
  , presets
  , decodeFn
  //, pauser
  , regexp  = '';

var mkdirp      = require('mkdirp')
  , moment      = require('moment')
  , request     = require('request')
  , path        = require('path')
  //, fs        = require('fs')
  , send        = require('send')
  //, pause      = require('pause')
  , crypto      = require('crypto')
  , lockFile    = require('lockfile')
  , gm          = require('gm')
  , sizeOf      = require('image-size')
  , fs          = require('fs-extra')
  , httpHandler = require('./http-handler')
  , mime        = require('mime');

var allowedMimeTypes = [
  "image/jpeg",
  "image/pjpeg",
  "image/gif",
  "image/png"
];

exports = module.exports = function (opts) {

  opts = opts || {};
  parseOptions(opts);

  if (opts.useIM) {
    gm = gm.subClass({imageMagick: true});
  }

  return function thumbs(req, res, next) {

    if ('GET' !== req.method && 'HEAD' !== req.method) { return next(); }
    //pauser = pause(req);

    function resume(runNext) {
      if (runNext) { next(); }
      //pauser.resume();
    }

    // Is this a request to a thumbnail image?
    var thumbRequestParts = req.originalUrl.match(regexp);
    var queryParams = req.query;

    if (!thumbRequestParts) {
      return resume(true);
    }

    var imagePreset = thumbRequestParts[1];

    if(!presets[imagePreset]){
      imagePreset = "";
    }

    var encodedImageURL = queryParams.base64;

    var hadError = false;
    var useFilesystem = true;
    if(!encodedImageURL){
      encodedImageURL = thumbRequestParts[2].replace(/(\?[a-zA-Z0-9&=]*)$/g, '');
      var originalDecodeFn = decodeFn;
      decodeFn = function(originalFilePath, callback){
        var ext = path.extname(thumbRequestParts[0]);
        callback(null, originalFilePath);
        decodeFn = originalDecodeFn;
      };
    }else{
      useFilesystem = false;
    }

    // Pre-declare variables that will be initialized in the decoder closure
    var filepath, fileStream, modifiedFilePath, preset;

    decodeFn(encodedImageURL, function imageURLDecoding(err, decodedImageURL) {
      //check if the decodedImageUrl's extension is good
      if(!isExtensionAllowed(decodedImageURL, opts)){
        res.writeHead(403);
        res.end('Forbidden File Extension!');
        return resume(false);
      }

      //-- Start creating and serving a thumbnail
      var targetDir = tmpDir + '/'+( imagePreset || "full" );
      var fullDir = tmpDir + '/full';

      // create the full directory
      mkdirp(fullDir, function (err) {
        if (err) {
          res.writeHead(500);
          res.end('tmpDir not writable!');
          console.log(err);
          return resume(false);
        }
      });

      mkdirp(targetDir, function (err) {
        if (err) {
          res.writeHead(500);
          res.end('tmpDir not writable!');
          console.log(err);
          return resume(false);
        }

        processFile();

      });

      function processFile() {
        var ext = path.extname(decodedImageURL);

        var hashedName = hash(req.originalUrl); // This is to be safe, in case somebody uses risky encodeFn
        var hashedOriginalName = hash(decodedImageURL); // This is to be safe, in case somebody uses risky encodeFn
        preset = (imagePreset ? "-" + imagePreset : "");
        filepath = fullDir + '/' + hashedOriginalName + ext;
        modifiedFilePath = targetDir + '/' + hashedName + preset + ext;

        var originalSize = {};
        try{
          originalSize = sizeOf(filepath);
        }catch(err){ }

        var options = presets[imagePreset] || {};
        try {
          if(queryParams.width || queryParams.w){
            options.width = parseInt(queryParams.width, 10) || parseInt(queryParams.w, 10);
          }
          if(queryParams.height || queryParams.h){
            options.height = parseInt(queryParams.height, 10) || parseInt(queryParams.h, 10);
          }
        }catch(err){}


        if(queryParams.quality || queryParams.q){
          options.quality = queryParams.quality || queryParams.q;
        }

        // obtain the size of an image if it exists and there are no options
        if(!options.width && !options.height) {
          options.width = originalSize.width;
          options.height = originalSize.height;
        }

        // see if we can serve the file from file cache, if ttl has not yet expired
        if (tmpCacheTTL > 0) {

          var modifiedFileSize = {};
          try{
            modifiedFileSize = sizeOf(modifiedFilePath);
          }catch(err){}

          try {
            var stats = fs.statSync(modifiedFilePath);
            var fileUnix = moment(stats.mtime).unix();
            var nowUnix = moment().unix();
            var diffUnix = nowUnix - fileUnix;
            if (diffUnix < tmpCacheTTL && modifiedFileSize.height === options.height && modifiedFileSize.width === options.width) {
              // file is fresh, no need to download/resize etc.
              var maxAge = ttl || 0;
              send(req, modifiedFilePath, {maxAge: maxAge})
                  .pipe(res);
              return resume(false);
            }
          } catch (err) {
            // no action necessary, just continue with normal flow
            console.log(err);
          }
        }

        var originalFileExists = false;
        try {
          var fileStats = fs.statSync(filepath);
          console.log("Using existing full file:", filepath);
          originalFileExists = true;
        }catch(err){}

        //create the file stream if it doesn't exist
        if(!originalFileExists){
          fileStream = fs.createWriteStream(filepath);
          processFileStream(fileStream, useFilesystem, options);
        }else{
          sendFileAfterTransform(options);
        }
      }

      function processFileStream(fileStream, useFilesystem, options){
        if(!useFilesystem){
          httpHandler.getFileHTTP(fileStream, request, decodedImageURL).then(function(){
            fileStream.on("close", function(){
              if(!hadError){
                sendFileAfterTransform(options);
              }else{
                fs.unlinkSync(filepath);
              }
            });
          }).catch(function(err){
            console.log(err);
            res.writeHead(404);
            res.end("Not Found!");
            hadError = true;
            return resume(false);
          });
        } else {
          fs.copy("app/images/"+decodedImageURL, filepath, function (err) {
            if (err){
              try{
                fs.unlinkSync(filepath);
              }catch(err){}

              res.writeHead(404);
              res.end("Not Found!");
              return resume(false);
            }else{
              sendFileAfterTransform(options);
            }
          });
        }
      }

      function sendFileAfterTransform(options) {
        var modificationOptions = {
          filepath: filepath
          , dstPath: modifiedFilePath
          , preset: options
          , smartCrop: opts.smartCrop || false
        };

        var postModificationFunction = function (err) {
          if (err) { throw err; }

          //console.log("SRC: " + req.originalUrl);
          //console.log("Streaming: " + modifiedFilePath);

          var maxTTLAge = ttl || 0;
          send(req, modifiedFilePath, {maxAge: maxTTLAge})
          //.on("error", function(err) {
          //  console.log("SENDING ERROR!");
          //  console.error(err);
          //})
          //  .on("end", function () { console.log("Streaming ended!"); })
              .pipe(res);

          return resume(false);
        };

        modifyImage(modificationOptions, postModificationFunction);

      }

    });
  };
};

exports.encodeURL = function(uri, callback) {
  callback(null, new Buffer(uri).toString('base64'));
};

exports.decodeURL = function(encodedURL, callback) {
  callback(null, new Buffer(encodedURL, 'base64').toString('ascii'));
};

/**
 * Return cryptographic hash (defaulting to: "sha1") of a string.
 *
 * @param {String} str
 * @param {String} algo - Algorithm used for hashing, defaults to sha1
 * @param {String} encoding - defaults to hex
 * @return {String}
 */
function hash(str, algo, encoding) {
  return crypto
    .createHash(algo || 'sha1')
    .update(str)
    .digest(encoding || 'hex');
}

/**
 * Merge user-provided options with the sensible defaults.
 * @param options
 */
function parseOptions(options) {

  ttl = options.ttl || (3600 * 24); // cache for 1 day by default.
  tmpCacheTTL = options.tmpCacheTTL || 5; // small by default
  decodeFn = options.decodeFn || exports.decodeURL;
  presets  = options.presets || defaultPresets();

  tmpDir   = options.tmpDir || '/tmp/nodethumbnails';

  var rootPath = options.rootPath || '/thumbs';
  if (rootPath[0] === '/') { rootPath = rootPath.substring(1); } // be forgiving to user errors!

  // Example: http://example.com/thumbs/small/AB23DC16Hash.jpg
  regexp = new RegExp('^\/' + rootPath.replace(/\//ig, '\\/') +
    '\/([A-Za-z0-9_]+)\/([%\.\-A-Za-z0-9_=\+?]*)', 'i');
}

function isExtensionAllowed(filePath, options){
  var allowedExtensions = options.allowedExtensions || ['gif', 'png', 'jpg', 'jpeg'];
  for (var i=0; i < allowedExtensions.length; i++) {
    // be forgiving to user errors!
    if (allowedExtensions[i][0] === '.') {
      allowedExtensions[i] = allowedExtensions[i].substring(1);
    }
  }

  var ext = filePath.split('.');
  ext = ext[ext.length - 1];

  return (allowedExtensions.indexOf(ext) > -1);
}

function defaultPresets() {

  return {
    small: {
      width: 120
      , quality: 50
    }
    , medium: {
      width: 300
      , quality: 70
    }
    , large: {
      width: 900
      , quality: 90
    }
  };

}

function modifyImage(options, callback) {
  var srcPath = options.filepath;
  var dstPath = options.dstPath;
  var preset = options.preset;

  var lockFileOptions = {
    wait: 1000,
    stale: 10000,
    retries: 3,
    retryWait: 250
  };

  var fileType = mime.lookup(srcPath);
  if(allowedMimeTypes.indexOf(fileType) > -1){
    gm(srcPath).identify(function (err, imageData) {
      if (err) { callback(err); }

      var origWidth  = imageData.size.width;
      var origHeight = imageData.size.height;

      if (!preset.quality) {
        preset.quality = 95;
      }

      if (options.smartCrop) {
        var Canvas = require('canvas'),
            SmartCrop = require('smartcrop');

        var img = new Canvas.Image();
        var canvasOptions = {}; canvasOptions.canvasFactory = function(w, h) { return new Canvas(w, h); };

        img.src = fs.readFileSync(srcPath);

        canvasOptions.width = preset.width;
        canvasOptions.height = preset.height || detectedHeight(canvasOptions.width, origWidth, origHeight);
        canvasOptions.dstPath = dstPath;

        SmartCrop.crop(img, canvasOptions, function(result) {
          //console.log(JSON.stringify(result.topCrop, null, '  '));
          var rt = result.topCrop;

          lockFile.lock(srcPath + ".lock", lockFileOptions, function(err) {
            if (err) { return callback(err); }
            gm(srcPath)
                .crop(rt.width, rt.height, rt.x, rt.y)
                .resize(preset.width || rt.width, preset.height || rt.height, "!")
                .gravity('Center')
                .quality(preset.quality)
                .write(dstPath, callback);

            lockFile.unlock(srcPath + ".lock", function (err) {
              if (err) { console.log(err); }
            });
          });
        });

      } else {

        var targetWidth = preset.width;
        // We could have just omitted height in .resize() but we may need control over this in the future.
        var targetHeight = preset.height || detectedHeight(targetWidth, origWidth, origHeight);
        //console.log ( "Target: " + targetWidth + " - " + targetHeight );

        lockFile.lock(srcPath + ".lock", lockFileOptions, function(err) {
          if (err) { return callback(err); }

          // @see: http://www.graphicsmagick.org/GraphicsMagick.html#details-resize
          gm(srcPath)
              .resize(targetWidth, targetHeight, "^")
              .crop(targetWidth, targetHeight, 0, 0)
              .quality(preset.quality)
              .write(dstPath, callback);

          lockFile.unlock(srcPath + ".lock", function (err) {
            if (err) { console.log(err); }
          });
        });
      }

    });
  }else{
    //unknown filetype
    res.writeHead(404);
    res.end("Not Found!");
    return resume(false);
  }
}

/**
 * Detect targetHeight for a proportional resizing, when only width is indicated.
 *
 * @param targetWidth
 * @param origWidth
 * @param origHeight
 */
function detectedHeight(targetWidth, origWidth, origHeight) {
  return origHeight * targetWidth / origWidth;
}
